File: demo/sorting/mergesort.py
def merge(left: list, right: list) -> list:
    """把两个分别排好序的序列归并成一个完整的排好序的序列

    Args:
        left (list): 参与归并的列表之一. 完整列表的"左半截", 已经按升序排好
        right (list): 参与归并的列表之一, 完整列表的"右半截", 已经按升序排好

    Returns:
        list: 归并好的列表, 整个按升序排列
    """    
    merged = [] # 准备把元素一个个append到这里, 以便返回
    while left or right: 
        # 如果把list的名字当作一个boolean型变量一样使用. 规则是：
        # list如果是空的, 没有元素, 就是False, 有元素就True
        if left and (not right or (right and left[0] < right[0])):
            # 从左半截的开头取元素, 需要两个条件同时满足：
            # 1. 左半截必须非空.
            # 2. 右半截要么空了, 要么开头的元素比左半截开头的元素更大.
            merged.append(left.pop(0)) # 从左边的开头取元素, 添加到归并好的序列的末尾
        else:
            merged.append(right.pop(0)) # 否则就从右边的开头取元素
    return merged


def mergesort(lst: list) -> list:
    '''把参数提供的list用归并排序排好, 并返回排好序的list (原未排好序的list不动)
    '''
    if len(lst) <= 1:
        return lst # 只有一个元素就不需要排序了
    else:
        # 两个整数之间/, 如果商是小数, 则返回小数; 但如果是//, 永远返回商的整数部分
        mid = len(lst) // 2
        # 复习：lst[a:b]是从a号开头到b-1号结尾之间的一段, 
        # a如果省略, 默认为0 (从头开始) ; b如果省略, 默认为len (到尾结束) 
        left = mergesort(lst[:mid]) # 把元素编号从0到 mid - 1 的左半段排好序
        right = mergesort(lst[mid:]) # 把右半段排好序
        return merge(left, right) # 归并左右两段并返回


def main():
    test_list = [8, 5, 2, 6, 4, 7, 1, 3]
    print(mergesort(test_list))
    test_list2 = [3, 8, 2, 6, 1, 9, 4, 2, 5]
    print(mergesort(test_list2))

if __name__ == "__main__":
    main()

File: demo/sorting/MergeSort.java
public class MergeSort {
    
  public static void mergeSort(int[] data, int start, int end) {
    // 注意：end 不作为“最后一个元素的编号”，而是最后一个元素的编号 + 1
    // 这样比较方便
    if (end - start > 1) { // start + 1 == end，是只有一个元素的base case，这时不需要做任何事
      int mid = start + (end - start) / 2; // 右半截的开头
      mergeSort(data, start, mid); // 把左半截排好序
      mergeSort(data, mid, end); // 把右半截排好序
      merge(data, start, mid, end); // 合并左右两截
    }
  }
    
  public static void merge(int[] data, int start, int mid, int end) {
    // 先新建两个array，分别存储左半截和右半截
    // 左右半截长度，如果Jessica考试时忘了结论，就找一组具体的数试一下
    int[] left = new int[mid - start];
    int[] right = new int[end - mid];
    // 把左右两个半截分别拷进去
    for (int i = 0; i < left.length; i++) {
      left[i] = data[i + start];
    }
    for (int i = 0; i < right.length; i++) {
      right[i] = data[i + mid];
    }
    // 准备好两个指针：
    int l = 0;
    int r = 0;
    // 直接把数据写回原本的 data 数组，所以还需要一个指针，决定写入位置
    int p = start;
    while (l < left.length || r < right.length) { // 只要两个半截的array没有都走完
      if ((l < left.length) && (r == right.length || left[l] <= right[r])) {
        // 从左半截拷一个元素的条件是：
        // 1. 左半截没有走完
        // 2. 要么右半截走完了，要么左边指针指的数字小于右边指针指的
        data[p] = left[l]; // 把左半截正在被指着的元素拷到原数组的待写入位置处
        l++; // 左半截的指针往右挪一格
      } else {
        // 只要两个半截没有都走完，而不从左边拷元素，就从右边拷
        data[p] = right[r];
        r++;
      }
      p++; // 无论从哪边拷元素，写入位置总往右挪一格
    }
  }
    
  public static void main(String[] args) {
    int[] arr = {8, 5, 2, 6, 4, 7, 1, 3};
    System.out.println("Before: ");
    for (int n : arr) {
      System.out.print(n);
      System.out.print('\t');
    }
    System.out.println("");
    mergeSort(arr, 0, arr.length);
    System.out.println("After: ");
    for (int n : arr) {
      System.out.print(n);
      System.out.print('\t');
    }
    System.out.println("");
  }
}